smart-contracts/contracts/cl-vault/src/rewards/distribution.rs

```diff
@@ -87,80 +116,119 @@ pub fn handle_collect_spread_rewards_reply(
     let mut rewards = CURRENT_REWARDS.load(deps.storage)?;
     rewards.update_rewards(response.collected_spread_rewards)?;
 
-    // update the rewards map against all user's locked up vault shares
-    distribute_rewards(deps, rewards)?;
+
+    // calculate and update the strategist fee
+    let vault_config = VAULT_CONFIG.load(deps.storage)?;
+    let strategist_fee = rewards.sub_ratio(vault_config.performance_fee)?;
+    STRATEGIST_REWARDS.update(deps.storage, |old| old.add(strategist_fee))?;
+
+    CURRENT_REWARDS.save(deps.storage, &rewards)?;
 
     // TODO add a nice response
     Ok(Response::new())
 }
 
-fn distribute_rewards(
-    mut deps: DepsMut,
-    mut rewards: CoinList,
-) -> Result<Vec<Attribute>, ContractError> {
-    if rewards.is_empty() {
-        return Ok(vec![Attribute::new("total_rewards_amount", "0")]);
+pub fn execute_distribute_rewards(
+    deps: DepsMut,
+    _env: Env,
+    amount_of_users: Uint128,
+) -> Result<Response, ContractError> {
+    let is_distributing = IS_DISTRIBUTING.may_load(deps.storage)?.unwrap_or(false);
+    if !is_distributing {
+        return Err(ContractError::IsNotDistributing {});
     }
 
-    let vault_config = VAULT_CONFIG.load(deps.storage)?;
-
-    // calculate the strategist fee
-    let strategist_fee = rewards.sub_ratio(vault_config.performance_fee)?;
-    STRATEGIST_REWARDS.update(deps.storage, |old| old.add(strategist_fee))?;
+    // Get current rewards to distribute, if there are not clear the state and return
+    let rewards = CURRENT_REWARDS.load(deps.storage)?;
+    deps.api.debug(format!("{:?}", rewards).as_str());
+    if rewards.is_empty() {
+        IS_DISTRIBUTING.save(deps.storage, &false)?;
+        return Ok(Response::new()
+            .add_attribute("total_rewards_amount", "0")
+            .add_attribute("is_last_distribution", "true"));
+    }
+    
+    let total_shares = CURRENT_TOTAL_SUPPLY.load(deps.storage)?;
+
+    let mut distributed_rewards = DISTRIBUTED_REWARDS
+        .load(deps.storage)
+        .unwrap_or(CoinList::new());
+
+    let mut users_processed: u128 = 0;
+
+    while users_processed < amount_of_users.u128() {
+        // Attempt to pop a (user_address, share_amount) tuple from the front of the snapshot Deque
+        if let Some((user_address, share_amount)) = DISTRIBUTION_SNAPSHOT.pop_front(deps.storage)? {
+            debug!(deps, "user_address", user_address);
+            debug!(deps, "share_amount", share_amount);
+
+            // Calculate the user's reward based on the share amount directly
+            let reward_ratio = Decimal::from_ratio(share_amount, total_shares.u128());
+            let user_reward = rewards.mul_ratio(reward_ratio);
+            let user_reward_clone = user_reward.clone();
```

Comment:
I assume there is some reason we have these two lines, but i'm not sure why.

Reply:
Yes, probably this could be optimized? I didn't manage to fix the borrowing issue right after, inside the USER_REWARDS.update() there is a user_reward_clone.add(old_rewards)?, that was trying to borrow the moved value.

Reply:
got it